src/xdsl/pattern_rewriter.py

```diff
@@ -0,0 +1,197 @@
+    """
+    # Get the operation argument and check that it is a subclass of Operation
+    params = [param for param in inspect.signature(func).parameters.values()]
+    if len(params) < 2:
+        raise Exception(
+            "op_type_rewrite_pattern expects the decorated function to "
+            "have two non-self arguments.")
+    is_method = params[0].name == "self"
+    if is_method:
+        if len(params) != 3:
+            raise Exception(
+                "op_type_rewrite_pattern expects the decorated method to "
+                "have two non-self arguments.")
+    else:
+        if len(params) != 2:
+            raise Exception(
+                "op_type_rewrite_pattern expects the decorated function to "
+                "have two arguments.")
+    expected_type = params[-2].annotation
+    if not issubclass(expected_type, Operation):
+        raise Exception(
+            "op_type_rewrite_pattern expects the first non-self argument"
+            "type hint to be an Operation subclass")
+
+    if not is_method:
+
+        def op_type_rewrite_pattern_static_wrapper(
+                op: Operation, rewriter: PatternRewriter) -> None:
+            if not isinstance(op, expected_type):
+                return None
+            func(op, rewriter)
+
+        return op_type_rewrite_pattern_static_wrapper
+
+    def op_type_rewrite_pattern_method_wrapper(
+            self, op: Operation, rewriter: PatternRewriter) -> None:
+        if not isinstance(op, expected_type):
+            return None
+        func(self, op, rewriter)
+
+    return op_type_rewrite_pattern_method_wrapper
+
+
+@dataclass(eq=False, repr=False)
+class GreedyRewritePatternApplier(RewritePattern):
+    """Apply a list of patterns in order until one pattern matches, and then use this rewrite."""
+
+    rewrite_patterns: List[RewritePattern]
+    """The list of rewrites to apply in order."""
+
+    def match_and_rewrite(self, op: Operation,
+                          rewriter: PatternRewriter) -> None:
+        for pattern in self.rewrite_patterns:
+            pattern.match_and_rewrite(op, rewriter)
+            if rewriter.has_done_action:
+                return
+        return
+
+
+@dataclass(eq=False, repr=False)
+class PatternRewriteWalker:
+    """
+    Walks the IR in the block and instruction order, and rewrite it in place.
+    Previous references to the walked operations are invalid after the walk.
+    Can walk either first the regions, or first the owner operation.
+    The walker will also walk recursively on the created operations.
+    """
+
+    pattern: RewritePattern
+    """Pattern to apply during the walk."""
+
+    walk_regions_first: bool = field(default=False)
+    """Choose if the walker should first walk the operation regions first, or the operation itself."""
+
+    apply_recursively: bool = field(default=True)
+    """Apply recursively rewrites on new operations."""
+
+    walk_reverse: bool = field(default=False)
```

Comment:
I can't really think of a case in which setting this actually matters.

Reply:
Assuming use cases while providing framework functionality is dangerous. If it is easy to implement, I'm happy with this extra feature.

Reply:
So in this case I added this because I needed it for my pass.
In this case, when walking in reverse, you are first reading the uses before the value definitions.
So the reason I need this is that contrary to the previous rewriter, when you delete an SSA value it deletes it everywhere. Here, when the value still has uses, it replaces them with a new SSA value called ErasedSSAValue. This is to make explicit that you are erasing it.
I'm starting to think that this might not be the best approach, though I'm not sure what to do here. One solution is to keep in ErasedSSAValue the old SSA value, and the other is to instead have a `erased` field in the SSAValue that get sets to true if we are erasing something that still has uses.